File: books/crafting_interpreters/rlox/lox/chunk.rs
use super::value;

#[derive(Debug)]
pub struct Chunk {
    pub code: Vec<OpCode>,
    pub constants: value::ValueArray,
    pub lines: Vec<usize>,
}

// TODO(Jonathon): Think about www.reddit.com/r/rust/comments/ui24sa/most_idiomatic_way_to_represent_opcodes/,
// and whether I should refactor this opcode setup.
#[derive(Debug)]
pub enum OpCode {
    OpConstant { constant: usize },
    OpAdd,
    OpSubtract,
    OpMultiply,
    OpDivide,
    OpNegate,
    OpReturn,
}

pub fn build_chunk() -> Chunk {
    Chunk {
        code: Vec::new(),
        constants: value::ValueArray::new(),
        lines: Vec::new(),
    }
}

pub fn write_chunk(chunk: &mut Chunk, code: OpCode, line: usize) {
    chunk.code.push(code);
    chunk.lines.push(line);
}

pub fn add_constant(chunk: &mut Chunk, value: value::Value) -> usize {
    value::write_value_array(&mut chunk.constants, value);
    return chunk.constants.len() - 1;
}

File: books/crafting_interpreters/pylox/lox/chunk.py
import dataclasses
import enum

import value


@enum.unique
class OpCode(enum.IntEnum):
    OP_CONSTANT = enum.auto()
    OP_RETURN = enum.auto()


@dataclasses.dataclass(frozen=False)
class Chunk:
    code: bytearray
    constants: list[value.Value]


def init_chunk() -> Chunk:
    return Chunk(
        code=bytearray(),
        constants=[],
    )


def write_chunk(chunk: Chunk, byte_: int) -> None:
    if 0 > byte_ > 255:
        raise ValueError(f"{byte_=} must be in the range [0-255].")
    chunk.code.append(byte_)


def add_constant(chunk: Chunk, val: value.Value) -> int:
    chunk.constants.append(val)
    return len(chunk.constants) - 1
